### ISSUE TITLE
Support Java 9

### ISSUE THREAD

**Issue by xuwei-k** [scala-js/scala-js#2798]
Here is reproduce example https://github.com/xuwei-k/scala-js-java9

- scala-js use ParTrieMap
  - https://github.com/scala-js/scala-js/blob/v0.6.14/tools/jvm/src/main/scala/org/scalajs/core/tools/linker/frontend/optimizer/ParIncOptimizer.scala#L36

- ParTrieMap call `scala.util.Properties.isJavaAtLeast`
  - https://github.com/scala/scala/blob/v2.10.6/src/library/scala/collection/parallel/package.scala#L44-L46
  - https://github.com/scala/scala/blob/v2.10.6/src/library/scala/collection/parallel/mutable/ParTrieMap.scala

- `scala.util.Properties.isJavaAtLeast` does not work with Scala 2.10.6, Java 9
  - fixed in Scala 2.12.1 https://github.com/scala/scala/pull/5276 

```
java.lang.ExceptionInInitializerError
	at scala.collection.parallel.ParIterableLike$class.$init$(ParIterableLike.scala:166)
	at scala.collection.parallel.mutable.ParTrieMap.<init>(ParTrieMap.scala:39)
	at scala.collection.parallel.mutable.ParTrieMap.<init>(ParTrieMap.scala:46)
	at scala.collection.parallel.mutable.ParTrieMap$.empty(ParTrieMap.scala:181)
	at org.scalajs.core.tools.linker.frontend.optimizer.ParIncOptimizer$CollOps$.emptyParMap(ParIncOptimizer.scala:36)
	at org.scalajs.core.tools.linker.frontend.optimizer.ParIncOptimizer$CollOps$.emptyParMap(ParIncOptimizer.scala:27)
	at org.scalajs.core.tools.linker.frontend.optimizer.GenIncOptimizer.<init>(GenIncOptimizer.scala:72)
	at org.scalajs.core.tools.linker.frontend.optimizer.ParIncOptimizer.<init>(ParIncOptimizer.scala:23)
	at org.scalajs.core.tools.linker.frontend.optimizer.ParIncOptimizer$$anonfun$1.apply(ParIncOptimizer.scala:204)
	at org.scalajs.core.tools.linker.frontend.optimizer.ParIncOptimizer$$anonfun$1.apply(ParIncOptimizer.scala:204)
	at org.scalajs.core.tools.linker.frontend.LinkerFrontend$$anonfun$1.apply(LinkerFrontend.scala:41)
	at org.scalajs.core.tools.linker.frontend.LinkerFrontend$$anonfun$1.apply(LinkerFrontend.scala:41)
	at scala.Option.map(Option.scala:145)
	at org.scalajs.core.tools.linker.frontend.LinkerFrontend.<init>(LinkerFrontend.scala:41)
	at org.scalajs.core.tools.linker.LinkerPlatformExtensions$class.apply(LinkerPlatformExtensions.scala:29)
	at org.scalajs.core.tools.linker.Linker$.apply(Linker.scala:78)
	at org.scalajs.sbtplugin.ScalaJSPluginInternal$$anonfun$org$scalajs$sbtplugin$ScalaJSPluginInternal$$scalaJSStageSettings$1$$anonfun$12.apply(ScalaJSPluginInternal.scala:207)
	at org.scalajs.sbtplugin.ScalaJSPluginInternal$$anonfun$org$scalajs$sbtplugin$ScalaJSPluginInternal$$scalaJSStageSettings$1$$anonfun$12.apply(ScalaJSPluginInternal.scala:206)
	at org.scalajs.core.tools.linker.ClearableLinker.org$scalajs$core$tools$linker$ClearableLinker$$ensureLinker(ClearableLinker.scala:78)
	at org.scalajs.core.tools.linker.ClearableLinker.semantics(ClearableLinker.scala:35)
	at org.scalajs.sbtplugin.ScalaJSPlugin$AutoImport$$anonfun$RhinoJSEnvInternal$1.apply(ScalaJSPlugin.scala:72)
	at org.scalajs.sbtplugin.ScalaJSPlugin$AutoImport$$anonfun$RhinoJSEnvInternal$1.apply(ScalaJSPlugin.scala:67)
	at scala.Function1$$anonfun$compose$1.apply(Function1.scala:47)
	at sbt.$tilde$greater$$anonfun$$u2219$1.apply(TypeFunctions.scala:40)
	at sbt.std.Transform$$anon$4.work(System.scala:63)
	at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
	at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
	at sbt.ErrorHandling$.wideConvert(ErrorHandling.scala:17)
	at sbt.Execute.work(Execute.scala:237)
	at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
	at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
	at sbt.ConcurrentRestrictions$$anon$4$$anonfun$1.apply(ConcurrentRestrictions.scala:159)
	at sbt.CompletionService$$anon$2.call(CompletionService.scala:28)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:514)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1161)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:844)
Caused by: java.lang.NumberFormatException: Not a version: 9
	at scala.util.PropertiesTrait$class.parts$1(Properties.scala:176)
	at scala.util.PropertiesTrait$class.isJavaAtLeast(Properties.scala:180)
	at scala.util.Properties$.isJavaAtLeast(Properties.scala:16)
	at scala.collection.parallel.package$.getTaskSupport(package.scala:45)
	at scala.collection.parallel.package$.<init>(package.scala:48)
	at scala.collection.parallel.package$.<clinit>(package.scala)
	at scala.collection.parallel.ParIterableLike$class.$init$(ParIterableLike.scala:166)
	at scala.collection.parallel.mutable.ParTrieMap.<init>(ParTrieMap.scala:39)
	at scala.collection.parallel.mutable.ParTrieMap.<init>(ParTrieMap.scala:46)
	at scala.collection.parallel.mutable.ParTrieMap$.empty(ParTrieMap.scala:181)
	at org.scalajs.core.tools.linker.frontend.optimizer.ParIncOptimizer$CollOps$.emptyParMap(ParIncOptimizer.scala:36)
	at org.scalajs.core.tools.linker.frontend.optimizer.ParIncOptimizer$CollOps$.emptyParMap(ParIncOptimizer.scala:27)
	at org.scalajs.core.tools.linker.frontend.optimizer.GenIncOptimizer.<init>(GenIncOptimizer.scala:72)
```

**Comment by sjrd** [comment]
Am I allowed to answer: this will be fixed when we use sbt 1.0, which uses Scala 2.12?

**Comment by sjrd** [comment]
Workaround: disable the parallel optimizer with:
```scala
scalaJSOptimizerOptions ~= { _.withParallel(false) }
```

**Comment by xuwei-k** [comment]
Thanks, it works :smiley: Should we change the default value like this?

```diff
--- a/sbt-plugin/src/main/scala/org/scalajs/sbtplugin/OptimizerOptions.scala
+++ b/sbt-plugin/src/main/scala/org/scalajs/sbtplugin/OptimizerOptions.scala
@@ -23,7 +23,15 @@ final class OptimizerOptions private (
     /** Whether to only warn if the linker has errors */
     val bypassLinkingErrors: Boolean = false,
     /** Whether to parallelize the optimizer (currently fastOptJS only) **/
-    val parallel: Boolean = true,
+    val parallel: Boolean = {
+      try {
+        scala.util.Properties.isJavaAtLeast("1.8")
+        true
+      } catch {
+        case _: NumberFormatException =>
+          false
+      }
+    },
```

**Comment by sjrd** [comment]
Hum, that's pretty ugly. But maybe that's the least ugly thing we can do to get this to work.

(At the very least it should be computed in a `private val` in the companion object, though.)

**Comment by sjrd** [comment]
Fixed in 5918016b88f420c9678888450e2cbe6095f4581d.

**Comment by OndraZizka** [comment]
Hi, I maintain a project, upgraded few scala related versions and got "Not a version: 9" during Maven build, without any context. I use SBT 1.1.0 and Scala 2.11.12. It's not clear to me from the comments above how to fix the issue. Any hint please?
